pkg/controller/snapshot_controller.go

```diff
@@ -421,13 +421,65 @@ func (ctrl *csiSnapshotController) checkandBindSnapshotContent(snapshot *crdv1.V
 	return nil
 }
 
+func (ctrl *csiSnapshotController) prepareCreateSnapshotInput(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshotClass, *v1.PersistentVolume, string, map[string]string, error) {
+	className := snapshot.Spec.VolumeSnapshotClassName
+	glog.V(5).Infof("prepareCreateSnapshotInput [%s]: VolumeSnapshotClassName [%s]", snapshot.Name, *className)
+	var class *crdv1.VolumeSnapshotClass
+	var err error
+	if className != nil {
+		class, err = ctrl.GetSnapshotClass(*className)
+		if err != nil {
+			glog.Errorf("prepareCreateSnapshotInput failed to getClassFromVolumeSnapshot %s", err)
+			return nil, nil, "", nil, err
+		}
+	} else {
+		glog.Errorf("failed to prepareCreateSnapshotInput %s without a snapshot class", snapshot.Name)
+		return nil, nil, "", nil, fmt.Errorf("failed to take snapshot %s without a snapshot class", snapshot.Name)
+	}
+
+	volume, err := ctrl.getVolumeFromVolumeSnapshot(snapshot)
+	if err != nil {
+		glog.Errorf("prepareCreateSnapshotInput failed to get PersistentVolume object [%s]: Error: [%#v]", snapshot.Name, err)
+		return nil, nil, "", nil, err
+	}
+
+	// Create VolumeSnapshotContent name
+	contentName := GetSnapshotContentNameForSnapshot(snapshot)
```

Comment:
maybe put this function outside of this function since it is not used when called by checkandUpdateBoundSnapshotStatusOperation?

maybe call prepareCreateSnapshotInput to getCreateSnapshotInfo?

Reply:
How about getCreateSnapshotInput?  "Info" would be confusing here because "Info" includes other information belonging to a snapshot.

Reply:
GetSnapshotContentNameForSnapshot is actually construct content name if using dynamic provisioning. Here if we use this function, statically binding case won't work? If snapshot is already bound, we can get reference name directly from snapshot object, right?

I am not familiar with the secrete ref function.  Why it needs contentName?

Reply:
contentName is used to construct secret namespace parameters. The secret is also needed at delete snapshot time when the snapshot object may not exist any more so we can't rely on the snapshot object.  The implementation to support secrets here is consistent with how it is done for persistent volumes.

I don't know if static binding can support secrets because snapshot class (and storage class) is not required for static binding so there may not be a way to pass in the secrets.  I'll need to do more investigation on that.

Reply:
How about this, change the logic to
if snapshot object already has a snapshotContentRef, use it, otherwise, construct the content name. Could you change the name from GetSnapshotContentNameForSnapshot to makeSnapshotContentName....?

Reply:
Sure.